### PR TITLE
Feature/enrich customer data tagging

### LINKED ISSUE #446: Enrich customer data tagging

### PR DESCRIPTION
**Pull request by olsi-qose**
## Description
Add new fields for customer tagging

## Related Issue
closes #446 

## Checklist:

- [x] My code follows the code style of this project.
- [x] I have assigned the correct milestone or created one if non existent.
- [x] I have correctly labeled this pull request.
- [x] I have linked the corresponding issue in this description.
- [x] I have updated the corresponding Jira ticket.
- [x] I have requested a review from at least 2 reviewers
- [x] I have checked the base branch of this pull request
- [x] I have checked my code for any possible security vulnerabilities

### CODE REVIEW

File: Model/Order/Buyer/Builder.php [hunk 1]
```diff
@@ -52,13 +52,18 @@ class Builder extends PersonBuilder
      * @inheritdoc
      * @return Buyer
      */
-    public function buildObject(
+    // phpcs:ignore Generic.CodeAnalysis.UnusedFunctionParameter.Found
```
Comment (supercid) [inline Model/Order/Buyer/Builder.php]: Remove this, the `// @codingStandardsIgnoreLine` on the line below should do the trick

File: Model/Person/Tagging/Builder.php [hunk 2]
```diff
@@ -37,57 +37,185 @@
 namespace Nosto\Tagging\Model\Person\Tagging;
 
 use Magento\Customer\Helper\Session\CurrentCustomer;
+use Magento\Framework\Event\ManagerInterface as EventManager;
+use Nosto\Nosto;
 use Nosto\Object\AbstractPerson;
```
Comment (supercid) [inline Model/Person/Tagging/Builder.php]: Seems to be unused

File: Model/Person/Tagging/Builder.php [hunk 3]
```diff
@@ -37,57 +37,185 @@
 namespace Nosto\Tagging\Model\Person\Tagging;
 
 use Magento\Customer\Helper\Session\CurrentCustomer;
+use Magento\Framework\Event\ManagerInterface as EventManager;
+use Nosto\Nosto;
```
Comment (supercid) [inline Model/Person/Tagging/Builder.php]: Seems to be unused

File: Model/Person/Tagging/Builder.php [hunk 4]
```diff
@@ -37,57 +37,185 @@
...
+     * @param CustomerRepositoryInterface $customerRepository
+     * @param NostoEmailRepository $emailRepository
+     * @param EventManager $eventManager
+     * @param NostoHelperData $nostoHelperData
+     */
+    public function __construct(
+        GroupRepository $groupRepository,
+        CustomerRepositoryInterface $customerRepository,
+        NostoEmailRepository $emailRepository,
+        NostoLogger $logger,
```
Comment (supercid) [inline Model/Person/Tagging/Builder.php]: Missing PHPDoc Arg

File: Model/Person/Tagging/Builder.php [hunk 5]
```diff
@@ -37,57 +37,185 @@
...
+        $customer->setLastName($lastName);
+        $customer->setEmail($email);
+        $customer->setPhone($phone);
+        $customer->setPostCode($postCode);
+        $customer->setCountry($country);
+        $customer->setCustomerGroup($customerGroup);
+        $customer->setCustomerReference($customerReference);
+        $customer->setGender($gender);
+        if ($dateOfBirth !== null) {
+            $customer->setDateOfBirth(\DateTime::createFromFormat("Y-m-d", $dateOfBirth));
```
Comment (supercid) [inline Model/Person/Tagging/Builder.php]: Single quotes since you are not evaluating an expression.

File: Model/Person/Tagging/Builder.php [hunk 6]
```diff
@@ -37,57 +37,185 @@
...
+
+        return $customer;
     }
 
     /**
      * Builds person from the current session / logged in user
      *
      * @param CurrentCustomer $currentCustomer
-     * @return AbstractPerson|null
+     * @return Customer|null
```
Comment (supercid) [inline Model/Person/Tagging/Builder.php]: The annotation on the builder function returns an object of type `AbstractPerson`. Please double check with the debugger and fix the annotations.

File: Model/Person/Tagging/Builder.php [hunk 7]
```diff
@@ -37,57 +37,185 @@
...
+
+    /**
+     * @param CustomerInterface $customer
+     * @return string
+     * @throws LocalizedException
+     * @throws NoSuchEntityException
+     */
+    private function getCustomerGroupName(CustomerInterface $customer)
+    {
+        $groupId = (int)$customer->getGroupId();
```
Comment (supercid) [inline Model/Person/Tagging/Builder.php]: `$customer->getGroupId()` can return null, add a nullcheck or `getById()` might break.

File: Model/Person/Tagging/Builder.php [hunk 8]
```diff
@@ -37,57 +37,185 @@
...
+    {
+        $customerReference = '';
+
+        try {
+            $customer = $currentCustomer->getCustomer();
+            $customerReference = $customer->getCustomAttribute(
+                NostoHelperData::NOSTO_CUSTOMER_REFERENCE_ATTRIBUTE_NAME
+            );
+
+            if (empty($customerReference)) {
```
Comment (supercid) [inline Model/Person/Tagging/Builder.php]: `$customer->getCustomAttribute` returns an object or null. Replace empty with `$customerReference === null`.

You can also invert the if and will look more readable.

File: Model/Person/Tagging/Builder.php (hunk 4 above)
Comment (hannunosto) [inline Model/Person/Tagging/Builder.php]: This is missing from the docblock

File: Model/Person/Tagging/Builder.php (hunk 3 above)
Comment (hannunosto) [inline Model/Person/Tagging/Builder.php]: This is not used

File: Model/Person/Tagging/Builder.php (hunk 2 above)
Comment (hannunosto) [inline Model/Person/Tagging/Builder.php]: This is not used

File: Model/Person/Tagging/Builder.php [hunk 9]
```diff
@@ -37,57 +37,187 @@
...
+     * @param CustomerInterface $customer
+     * @return string
+     * @throws LocalizedException
+     * @throws NoSuchEntityException
+     */
+    private function getCustomerGroupName(CustomerInterface $customer)
+    {
+        $groupId = (int)$customer->getGroupId();
+        if ($groupId === null) {
+            return '';
```
Comment (hannunosto) [inline Model/Person/Tagging/Builder.php]: Let's return `null`

File: Model/Person/Builder.php [hunk 10]
```diff
@@ -89,7 +89,11 @@ public function build(
         $email,
         $phone = null,
         $postCode = null,
-        $country = null
+        $country = null,
+        $customerGroup = null,
```
Comment (supercid) [inline Model/Person/Builder.php]: They're all missing the docblock

**Pull request merged**